### ISSUE TITLE
How does a 5 category problem become a 3 category problem？

### ISSUE THREAD

**Issue by quuhua911** [awslabs/autogluon#201]
I did a Tabular Prediction test。It was originally a five-category forecast problem，but AutoGluon only keep 3 out of 5 classes？How do I solve this problem by adding parameters?

predictor = task.fit(train_data=train_data, label=label_column,problem_type="multiclass", output_directory=dir)

Beginning AutoGluon training ...
Preprocessing data ...
Warning: Some classes in data were not valid. AutoGluon will 
only keep 3 out of 5 classes

Fraction of data that will be kept for training models: 0.972
	Data preprocessing and feature engineering runtime = 0.11s ...
AutoGluon will gauge predictive performance using evaluation metric: accuracy
To change this, specify the eval_metric argument of fit()
Fitting model: RandomForestClassifierGini ...
	0.39s	 = Training runtime
	0.9694	 = Validation accuracy score
Fitting model: RandomForestClassifierEntr ...
	0.42s	 = Training runtime
	0.9592	 = Validation accuracy score
Fitting model: ExtraTreesClassifierGini ...
	0.3s	 = Training runtime
	0.9694	 = Validation accuracy score
Fitting model: ExtraTreesClassifierEntr ...
	0.3s	 = Training runtime
	0.9694	 = Validation accuracy score
Fitting model: KNeighborsClassifierUnif ...
	0.01s	 = Training runtime
	0.898	 = Validation accuracy score

**Comment by Innixma** [comment]
```
label_count_threshold : int, default = 10
                For multi-class classification problems, this is the minimum number of times a label must appear in dataset in order to be considered an output class. 
                AutoGluon will ignore any classes whose labels do not appear at least this many times in the dataset (i.e. will never predict them). 
```

The logging message is vague at the moment, I will look towards improving it so it makes more sense. This is occurring because your dataset has very few instances of the classes in question, and so it is difficult to do a reasonable train/validation split on the data and get enough data to make meaningful predictions on those classes.

I would recommend increasing the size of your data to have enough examples or keeping things as they are with the dropped classes. Alternatively, set the threshold to a lower value.

**Comment by jwmueller** [comment]
@quuhua911  This happens because Autogluon will ignore any classes that do not appear at least label_count_threshold times in the dataset (default = 10), which means the fitted AutoGluon models will never predict such a class.  With such a rarely-occurring class, AutoGluon models are unlikely to learn how to predict it accurately in the first place, even if it were not ignored.  Note that just because AutoGluon models never predict this rare class, this does not mean they will perform poorly and test-accuracy of the models can still be computed with all of the existing classes.

If you really want to include such a rare class so it sometimes gets predicted, you can decrease `label_count_threshold`, or you can also duplicate the datapoints from this rare class in the training data so that the resulting fitted models are encouraged to predict this class more often.

**Comment by quuhua911** [comment]
> ```
> label_count_threshold : int, default = 10
>                 For multi-class classification problems, this is the minimum number of times a label must appear in dataset in order to be considered an output class. 
>                 AutoGluon will ignore any classes whose labels do not appear at least this many times in the dataset (i.e. will never predict them). 
> ```
> 
> The logging message is vague at the moment, I will look towards improving it so it makes more sense. This is occurring because your dataset has very few instances of the classes in question, and so it is difficult to do a reasonable train/validation split on the data and get enough data to make meaningful predictions on those classes.
> 
> I would recommend increasing the size of your data to have enough examples or keeping things as they are with the dropped classes. Alternatively, set the threshold to a lower value.

---
The training set I used, the smallest category, has a sample size of 2293。The details are as follows。So There seems to be a bug in the current program。Besides，I set label_count_threshold=1 ,The training program fit() can run, but there are only four categories of predictions result.
The last question is, what does the parameter "label_count_threshold" mean? Does it represent the number of samples or the percentage?

all_data_train["label"].value_counts().sort_index(ascending=False)
C7E2941B65C6CCD6      2293
816A9BEBED2D7C99     17812
7C26FADD409BD4B9    282906
56AFA2A526F96CC9      8954
0F2E4CC10EDBE80F     10792
Name: label, dtype: int64

**Comment by quuhua911** [comment]
The training set I used, the smallest category, has a sample size of 2293。The details are as follows。So There seems to be a bug in the current program。Besides，I set label_count_threshold=1 ,The training program fit() can run, but there are only four categories of predictions result.
The last question is, what does the parameter "label_count_threshold" mean? Does it represent the number of samples or the percentage?


all_data_train["label"].value_counts().sort_index(ascending=False)
C7E2941B65C6CCD6 2293
816A9BEBED2D7C99 17812
7C26FADD409BD4B9 282906
56AFA2A526F96CC9 8954
0F2E4CC10EDBE80F 10792
Name: label, dtype: int64





The email attachment is my training set and test set，The＂ label ＂column is the target column that I'm going to predict．
Data set mixed with a bit of Chinese characters, the original data set is so, if the program can not deal with chinese characters,you can directly delete these result, this should not greatly affect the prediction results.
After I deleted earlier, using autogluon still didn't predict correctly.


Thank you.









------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Jonas Mueller"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月15日(星期三) 凌晨2:12
收件人:&nbsp;"awslabs/autogluon"<autogluon@noreply.github.com&gt;;
抄送:&nbsp;"云卷云舒"<20613622@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [awslabs/autogluon] How does a 5 category problem become a 3 category problem？ (#201)




@quuhua911  This happens because Autogluon will ignore any classes that do not appear at least label_count_threshold times in the dataset (default = 10), which means the fitted AutoGluon models will never predict such a class.  With such a rarely-occurring class, AutoGluon models are unlikely to learn how to predict it accurately in the first place, even if it were not ignored.  Note that just because AutoGluon models never predict this rare class, this does not mean they will perform poorly and test-accuracy of the models can still be computed with all of the existing classes.
 
If you really want to include such a rare class so it sometimes gets predicted, you can decrease label_count_threshold, or you can also duplicate the datapoints from this rare class in the training data so that the resulting fitted models are encouraged to predict this class more often.

**Comment by quuhua911** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"20613622"<20613622@qq.com&gt;;
发送时间:&nbsp;2020年1月15日(星期三) 中午11:19
收件人:&nbsp;"awslabs/autogluon"<reply@reply.github.com&gt;;

主题:&nbsp;回复： [awslabs/autogluon] How does a 5 category problem become a 3 category problem？ (#201)





The training set I used, the smallest category, has a sample size of 2293。The details are as follows。So There seems to be a bug in the current program。Besides，I set label_count_threshold=1 ,The training program fit() can run, but there are only four categories of predictions result.
The last question is, what does the parameter "label_count_threshold" mean? Does it represent the number of samples or the percentage?


all_data_train["label"].value_counts().sort_index(ascending=False)
C7E2941B65C6CCD6 2293
816A9BEBED2D7C99 17812
7C26FADD409BD4B9 282906
56AFA2A526F96CC9 8954
0F2E4CC10EDBE80F 10792
Name: label, dtype: int64





The email attachment is my training set and test set，The＂ label ＂column is the target column that I'm going to predict．
Data set mixed with a bit of Chinese characters, the original data set is so, if the program can not deal with chinese characters,you can directly delete these result, this should not greatly affect the prediction results.
After I deleted earlier, using autogluon still didn't predict correctly.


Thank you.









------------------ 原始邮件 ------------------
发件人:&nbsp;"Jonas Mueller"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月15日(星期三) 凌晨2:12
收件人:&nbsp;"awslabs/autogluon"<autogluon@noreply.github.com&gt;;
抄送:&nbsp;"云卷云舒"<20613622@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [awslabs/autogluon] How does a 5 category problem become a 3 category problem？ (#201)




@quuhua911  This happens because Autogluon will ignore any classes that do not appear at least label_count_threshold times in the dataset (default = 10), which means the fitted AutoGluon models will never predict such a class.  With such a rarely-occurring class, AutoGluon models are unlikely to learn how to predict it accurately in the first place, even if it were not ignored.  Note that just because AutoGluon models never predict this rare class, this does not mean they will perform poorly and test-accuracy of the models can still be computed with all of the existing classes.
 
If you really want to include such a rare class so it sometimes gets predicted, you can decrease label_count_threshold, or you can also duplicate the datapoints from this rare class in the training data so that the resulting fitted models are encouraged to predict this class more often.
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.



从QQ邮箱发来的超大附件

all_data_train.csv (114.69M, 2020年02月14日 11:27 到期)进入下载页面：http://mail.qq.com/cgi-bin/ftnExs_download?t=exs_ftn_download&k=746566355bdab6c1c5ec5c3417345016555700050651030f1e015f510719560001074b570652521403545550010c030a01520707312662585f093951504003664717075c5f1a014a45655b&code=3ef514b9




all_data_test.csv (28.00M, 2020年01月15日 11:31 到期)进入下载页面：http://mail.qq.com/cgi-bin/ftnExs_download?t=exs_ftn_download&k=2063313779cc06ca90ea0b3616360b1b5e57030f08075a0c4b01550e011b0d0c05501c56005458195550060f06570000075b0354302739550a0f6e535142586b120642431e554a42665e&code=fc170694

**Comment by Innixma** [comment]
Thanks for the additional information!

We just released AutoGluon 0.0.4 which contains a host of bug fixes. Can you try the new version and see if your issue is resolved?

Best,
Nick

**Comment by Innixma** [comment]
@quuhua911 Has the problem been resolved? Please let us know!

**Comment by quuhua911** [comment]
the problem has  been resolved。Thank you. But I checked the results, autogluon is still less accurate than manual tuning.

**Comment by quuhua911** [comment]
I mean, the training would have been better if the raw data had been manually culled for the outliers, Data pre-processing, and then autogulon.

**Comment by Innixma** [comment]
Thanks for confirming that the problem has been resolved.

AutoGluon can frequently benefit from manual feature engineering and pre-processing prior to sending to the .fit call. The scope of different problems within Tabular is extremely vast, so it is difficult to always perform the best pre-processing automatically.

If you want to improve results, try setting auto_stack=True, this should get better scores.

If you have ideas for generic pre-processing steps that could improve AutoGluon, please feel free to elaborate in a new issue or submit a PR. For now I will mark this as resolved.

Best,
Nick
